hrms/payroll/doctype/salary_slip/salary_slip.py

```diff
@@ -70,12 +72,40 @@ def __init__(self, *args, **kwargs):
 	def autoname(self):
 		self.name = make_autoname(self.series)
 
+	@property
+	def joining_date(self):
+		if not hasattr(self, "__joining_date"):
+			self.__joining_date = frappe.get_cached_value(
+				"Employee",
+				self.employee,
+				"date_of_joining",
+			)
+
+		return self.__joining_date
+
+	@property
+	def relieving_date(self):
+		if not hasattr(self, "__relieving_date"):
+			self.__relieving_date = frappe.get_cached_value(
+				"Employee",
+				self.employee,
+				"relieving_date",
+			)
+
+		return self.__relieving_date
+
 	@property
 	def payroll_period(self):
-		if not hasattr(self, "_payroll_period"):
-			self._payroll_period = get_payroll_period(self.start_date, self.end_date, self.company)
+		if not hasattr(self, "__payroll_period"):
+			if frappe.flags.in_test:
+				self.__payroll_period = get_payroll_period(self.start_date, self.end_date, self.company)
+			else:
+				self.__payroll_period = frappe.cache().hget("payroll_period", "payroll_period")
+				if not self.__payroll_period:
+					self.__payroll_period = get_payroll_period(self.start_date, self.end_date, self.company)
+					frappe.cache().hset("payroll_period", "payroll_period", self.__payroll_period)
```

Comment:
@saurabh6790 

This will lead to incorrect results:
- Run payroll for company A. Fetches the correct payroll period. Uses cached value in subsequent requests ✅
- Run payroll for company B. The first payroll period is still in the cache, might end up using the incorrect payroll period

Solution: Use unique keys with company & dates. This might be too much. Instead, I think memoizing this function call with @redis_cache will be better. The default TTL is 60 mins. Will also add cache invalidation on payroll period updates.

Let me know what you think